heclib/hecdss/hecdss.c

```diff
@@ -0,0 +1,742 @@
+
+  if (saveAsFloat) {
+    float* values = hec_dss_double_array_to_float(valueArray, valueArraySize);
+    if (values == 0)
+      return -1;
+
+    tss = zstructTsNewRegFloats(pathname, values, valueArraySize, startDate, startTime, units, type);
+    tss->allocated[zSTRUCT_TS_floatValues];// zstructFree will free float array
+    if (qualityArraySize > 0 && qualityArraySize == valueArraySize)
+      tss->quality = qualityArray;
+  }
+  else {
+    tss = zstructTsNewRegDoubles(pathname, valueArray, valueArraySize, startDate, startTime, units, type);
+  }
+  
+  const int storageFlag = 0;// storageFlag = 0  Always replace data (Regular Interval)
+  //	 storageFlag = 1  replace  (Irregular)
+
+  int status = ztsStore(dss->ifltab, tss, storageFlag);
+  
+
+  zstructFree(tss);
+  return status;
+}
+HECDSS_API int hec_dss_tsStoreIregular(dss_file* dss, const char* pathname,
+  const char* startDateBase,
+  int* times,const int timeGranularitySeconds,
+  double* valueArray, const int valueArraySize,
+  int* qualityArray, const int qualityArraySize,
+  const int saveAsFloat,
+  const char* units, const char* type)
+{
+  zStructTimeSeries* tss = 0;
+
+  if (saveAsFloat) {
+    float* values = hec_dss_double_array_to_float(valueArray, valueArraySize);
+    if (values == 0) {
+      return -1;
+    }
+
+    tss = zstructTsNewIrregFloats(pathname,values, valueArraySize,times,
+      timeGranularitySeconds, startDateBase,units, type);
+    tss->allocated[zSTRUCT_TS_floatValues];
+    if (qualityArraySize > 0 && qualityArraySize == valueArraySize)
+      tss->quality = qualityArray;
+  }
+  else {
+    tss = zstructTsNewIrregDoubles(pathname, valueArray, valueArraySize, times,
+      timeGranularitySeconds, startDateBase, units, type);
+  }
+
+  const int storageFlag = 1;// storageFlag = 0  Always replace data (Regular Interval)
+  //	 storageFlag = 1  replace  (Irregular)
+
+  int status = ztsStore(dss->ifltab, tss, storageFlag);
+
+
+  zstructFree(tss);
+  return status;
+}
+
+HECDSS_API int hec_dss_locationRetrieve(dss_file* dss, const char* fullPath, 
+                            double* x,double* y, double* z,
+                            int* coordinateSystem, int* coordinateID, 
+                            int* horizontalUnits,int* horizontalDatum,
+                            int* verticalUnits, int* verticalDatum,
+                            char* timeZoneName, const int timeZoneNameLength,
+                            char* supplemental, const int supplementalLength){
+
+  zStructLocation* loc = zstructLocationNew(fullPath);
+  int status = zlocationRetrieve(dss->ifltab, loc);
+  if (status == 0) {
+    *x = loc->xOrdinate;
+    *y = loc->yOrdinate;
+    *z = loc->zOrdinate;
+    *coordinateSystem = loc->coordinateSystem;
+    *coordinateID = loc->coordinateID;
+    *verticalUnits = loc->verticalUnits;
+    *verticalDatum = loc->verticalDatum;
+    *horizontalUnits = loc->horizontalUnits;
+    *horizontalDatum = loc->horizontalDatum;
+
```

Comment:
Initialize char* out parameters below to empty strings either before `stringCopy(...)` or in `else` clauses.